asterius/src/Asterius/Types/SymbolSet.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeFamilies #-}
+
+-- |
+-- Module      :  Asterius.Types.SymbolSet
+-- Copyright   :  (c) 2018 EURL Tweag
+-- License     :  All rights reserved (see LICENCE file in the distribution).
+--
+-- The 'SymbolSet' type represents a set of elements of type 'EntitySymbol'.
+-- Internally it is represented as an 'IM.IntMap', mapping the keys of the
+-- uniques of the 'EntitySymbol's to the 'EntitySymbol's themselves.
+module Asterius.Types.SymbolSet
+  ( -- * SymbolSet type
+    SymbolSet,
+
+    -- * Construction
+    empty,
+    singleton,
+
+    -- * Query
+    null,
+
+    -- * Combine
+    union,
+    unions,
+    difference,
+
+    -- * Folds and Maps
+    foldr',
+
+    -- * Conversion
+    toIntSet,
+
+    -- ** Lists
+    fromList,
+    toList,
+  )
+where
+
+import Asterius.Types.EntitySymbol
+import Data.Data
+import qualified Data.Foldable as Foldable
+import qualified Data.IntMap.Lazy as IM
+import qualified Data.IntSet as IS
+import GHC.Exts (IsList (..))
+import Prelude hiding (null)
+
+-- | A set of 'EntitySymbol's.
+newtype SymbolSet = SymbolSet {fromSymbolSet :: IM.IntMap EntitySymbol}
+  deriving newtype (Eq, Semigroup, Monoid)
+  deriving stock (Data)
+
+instance Show SymbolSet where
+  showsPrec p (SymbolSet s) =
+    showParen (p > 10) $
+      showString "fromList " . shows (toList s)
+
+instance IsList SymbolSet where
+  type Item SymbolSet = EntitySymbol
+  fromList = fromListSS
+  toList = toListSS
+
+-- ----------------------------------------------------------------------------
+
+-- | /O(1)/. The empty set.
+{-# INLINE empty #-}
+empty :: SymbolSet
+empty = SymbolSet IM.empty
+
+-- | /O(1)/. A set of one element.
+{-# INLINE singleton #-}
+singleton :: EntitySymbol -> SymbolSet
+singleton e = SymbolSet $ IM.singleton (getKeyES e) e
+
+-- | /O(n+m)/. The union of two sets.
+{-# INLINE union #-}
+union :: SymbolSet -> SymbolSet -> SymbolSet
```

Comment:
I think we can directly `coerce IM.union` here.

Reply:
The same may apply to all other functions which simply wraps/unwraps the `newtype` around the `IntMap` and reuses an `IntMap` interface. Always prefer `coerce` if possible.

Reply:
I did apply your advice to all functions that I could in this file (though I needed to add some explicit type applications to bypass GHC's failing type inference). 

> Always prefer `coerce` if possible.

I see that with `coerce` the inliner gets to inline the functions more eagerly so that's nice. Is there any other benefit in choosing `coerce` over newtype wrapping/unwrapping? I have the impression that it's all the same to GHC; such no-ops are eventually optimized anyway.

Reply:
IMO it's mainly about code brevity.